### PR TITLE
[SPARK-37270][SQL] Fix push foldable into CaseWhen branches if elseValue is empty

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr fix push foldable into CaseWhen branches if elseValue is empty. For example:
```scala
spark.sql("CREATE TABLE t1(bool boolean, number int) using parquet")
spark.sql("INSERT INTO t1 VALUES(false, 1)")

spark.sql(
  """
    |SELECT *
    |FROM   (SELECT *,
    |               CASE
    |                 WHEN bool THEN 'I am not null'
    |               END AS conditions
    |        FROM   t1) t
    |WHERE  conditions IS NULL
    |""".stripMargin).show
```

How do we optimize the filter conditions before this pr:
```
Filter isnull(CASE WHEN bool#7 THEN I am not null END) -> 
Filter CASE WHEN bool#7 THEN isnull(I am not null) END -> // After PushFoldableIntoBranches
Filter (bool#7 AND isnull(I am not null)) -> 
Filter (bool#7 AND false) ->
Filter false
```

How do we optimize the filter conditions after this pr:
```
Filter isnull(CASE WHEN bool#7 THEN I am not null END) -> 
Filter CASE WHEN bool#7 THEN isnull(I am not null) ELSE isnull(null) END -> // After PushFoldableIntoBranches
Filter CASE WHEN bool#7 THEN false ELSE isnull(null) END -> 
Filter CASE WHEN bool#7 THEN false ELSE true END -> 
Filter NOT (bool#7 <=> true)
```

### Why are the changes needed?

Fix correctness issue.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -681,7 +681,9 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
           if supportedUnaryExpression(u) && atMostOneUnfoldable(branches.map(_._2) ++ elseValue) =>
         c.copy(
           branches.map(e => e.copy(_2 = u.withNewChildren(Array(e._2)))),
-          elseValue.map(e => u.withNewChildren(Array(e))))
+          elseValue
+            .orElse { if (u.isInstanceOf[IsNull]) Some(Literal(null, c.dataType)) else None }
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:685]: Wouldn't it be safer to allways return `Some(Literal(null, c.dataType))` here?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:685]: Yes. It should be safer, but it will add a useless `Literal(null, c.dataType)` in most cases. Another way is:
```
elseValue.orElse(if (u.isInstanceOf[Predicate]) Some(Literal(null, c.dataType)) else None)
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:685]: Can we optimize away the useless else branch? I think null propagation + removing null else branch should do the work?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala [hunk 2]
```diff
@@ -356,4 +356,27 @@ class PushFoldableIntoBranchesSuite
...
       Not('a > 10 <=> TrueLiteral))
   }
+
+  test("SPARK-37270: Fix push foldable into CaseWhen branches if elseValue is empty") {
+    assertEquivalent(
+      IsNull(CaseWhen(Seq(('a > 10, Literal(0))), Literal(1))),
+      FalseLiteral)
+    assertEquivalent(
+      IsNull(CaseWhen(Seq(('a > 10, Literal(0))))),
+      !('a > 10 <=> true))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala:366]: not related to this PR: do we have rules to optimize `a = true` to `a`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala [hunk 3]
```diff
@@ -356,4 +356,27 @@ class PushFoldableIntoBranchesSuite
...
+
+    assertEquivalent(
+      CaseWhen(Seq(('a > 10, Literal(0))), Literal(1)) <=> Literal(null, IntegerType),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala:369]: how is this related to `PushFoldableIntoBranches`?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala:369]: This is an example:
```scala
spark.sql("CREATE TABLE t1(bool boolean, number int) using parquet")
spark.sql("INSERT INTO t1 VALUES(false, 1)")

spark.sql(
  """
    |SELECT *
    |FROM   (SELECT *,
    |               CASE
    |                 WHEN bool THEN 'I am not null'
    |               END AS conditions
    |        FROM   t1) t
    |WHERE  conditions <=> NULL
    |""".stripMargin).show
```


Before this pr:
```
Filter isnull(CASE WHEN bool#7 THEN I am not null END) ->
Filter CASE WHEN bool#7 THEN isnull(I am not null) END -> // After PushFoldableIntoBranches
Filter (bool#7 AND isnull(I am not null)) ->
Filter (bool#7 AND false) ->
Filter false
```

After this pr:
```
Filter isnull(CASE WHEN bool#7 THEN I am not null END) -> 
Filter CASE WHEN bool#7 THEN isnull(I am not null) ELSE isnull(null) END -> // After PushFoldableIntoBranches
Filter CASE WHEN bool#7 THEN false ELSE isnull(null) END -> 
Filter CASE WHEN bool#7 THEN false ELSE true END -> 
Filter NOT (bool#7 <=> true)
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicateSuite.scala [hunk 4]
```diff
@@ -108,15 +108,18 @@ class SimplifyConditionalsInPredicateSuite extends PlanTest {
     Seq(Some(FalseLiteral), None, Some(Literal(null, BooleanType))).foreach { elseExp =>
       val originalCond = CaseWhen(
         Seq((UnresolvedAttribute("i") > Literal(10), UnresolvedAttribute("b"))),
-        elseExp)
+        elseExp).as("c")
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicateSuite.scala:111]: Add alias to avoid test failure:
```
== FAIL: Plans do not match ===
!Project [CASE WHEN (i#0 > 10) THEN b#0 END AS CASE WHEN (i > 10) THEN b ELSE NULL END#0]   Project [CASE WHEN (i#0 > 10) THEN b#0 END AS CASE WHEN (i > 10) THEN b END#0]
 +- LocalRelation <empty>, [i#0, b#0, a#0, m#0]                                             +- LocalRelation <empty>, [i#0, b#0, a#0, m#0]

```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala:369]: Does this two UT fail before your PR?

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-modified/q34.sf100/explain.txt [hunk 5]
```diff
@@ -98,15 +98,15 @@ Input [5]: [ss_customer_sk#1, ss_hdemo_sk#2, ss_store_sk#3, ss_ticket_number#4,
 Output [4]: [hd_demo_sk#11, hd_buy_potential#12, hd_dep_count#13, hd_vehicle_count#14]
 Batched: true
 Location [not included in comparison]/{warehouse_dir}/household_demographics]
-PushedFilters: [IsNotNull(hd_vehicle_count), IsNotNull(hd_dep_count), Or(EqualTo(hd_buy_potential,>10000         ),EqualTo(hd_buy_potential,Unknown        )), GreaterThan(hd_vehicle_count,0), IsNotNull(hd_demo_sk)]
```
Comment (cloud-fan) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-modified/q34.sf100/explain.txt:101]: hmm, we optimized away `IsNotNull(hd_dep_count)`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala:369]: I don't understand how can `PushFoldableIntoBranches` transform these 2 `CaseWhen`.

Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala:369]: Yes. These two UT will fail. I have Updated the last comment:  https://github.com/apache/spark/pull/34580#discussion_r749259469

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-modified/q34.sf100/explain.txt (hunk 5 above)
Comment (wangyum) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-modified/q34.sf100/explain.txt]: This is because we remove the `elseValue` if it is null. Fixed by https://github.com/apache/spark/pull/34580/commits/26436d27384771240d7d4e12c85f11ce67cee5a8

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicateSuite.scala [hunk 6]
```diff
@@ -222,7 +225,7 @@ class SimplifyConditionalsInPredicateSuite extends PlanTest {
   }
 
   private def testProjection(originalExpr: Expression, expectedExpr: Expression): Unit = {
-    test((rel, exp) => rel.select(exp), originalExpr, expectedExpr)
+    test((rel, exp) => rel.select(exp), originalExpr.as("out"), expectedExpr.as("out"))
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicateSuite.scala:228]: Add alias to avoid test failure:
```
== FAIL: Plans do not match ===
!Project [CASE WHEN (i#0 > 10) THEN b#0 END AS CASE WHEN (i > 10) THEN b ELSE NULL END#0]   Project [CASE WHEN (i#0 > 10) THEN b#0 END AS CASE WHEN (i > 10) THEN b END#0]
 +- LocalRelation <empty>, [i#0, b#0, a#0, m#0]                                             +- LocalRelation <empty>, [i#0, b#0, a#0, m#0]

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 7]
```diff
@@ -678,10 +684,11 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
           falseValue = u.withNewChildren(Array(falseValue)))
 
       case u @ UnaryExpression(c @ CaseWhen(branches, elseValue))
-          if supportedUnaryExpression(u) && atMostOneUnfoldable(branches.map(_._2) ++ elseValue) =>
+          if supportedUnaryExpression(u) && atMostOneUnfoldable(branches.map(_._2) :+
+            elseValue.getOrElse(Literal(null, c.dataType))) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:688]: If the CaseWhen has only one branch and has no else branch, we can still pushdown?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:688]: Yes. It can push down more filters. For example:
```scala
spark.sql("CREATE TABLE t1(a int, b int) using parquet")
spark.sql("SELECT * FROM t1 WHERE CASE WHEN a > 10 THEN b END = 20").explain
```
Pushdown:
```
== Physical Plan ==
*(1) Filter (((isnotnull(a#0) AND isnotnull(b#1)) AND (a#0 > 10)) AND (b#1 = 20))
+- *(1) ColumnarToRow
   +- FileScan parquet default.t1[a#0,b#1] Batched: true, DataFilters: [isnotnull(a#0), isnotnull(b#1), (a#0 > 10), (b#1 = 20)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/SPARK-31890/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(a), IsNotNull(b), GreaterThan(a,10), EqualTo(b,20)], ReadSchema: struct<a:int,b:int>
```
Do not pushdown:
```
== Physical Plan ==
*(1) Filter (CASE WHEN (a#0 > 10) THEN b#1 END = 20)
+- *(1) ColumnarToRow
   +- FileScan parquet default.t1[a#0,b#1] Batched: true, DataFilters: [(CASE WHEN (a#0 > 10) THEN b#1 END = 20)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/SPARK-31890/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:int,b:int>
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:688]: Got it. Then we should probably update `atMostOneUnfoldable`
```
  private def atMostOneUnfoldable(exprs: Seq[Expression]): Boolean = {
    if (exprs.length < 2) return true
    val (foldables, others) = exprs.partition(_.foldable)
    foldables.nonEmpty && others.length < 2
  }
```

Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:688]: Isn't this equivalent to just removing the `foldables.nonEmpty` condition?

**Comment by wangyum** [comment]
Make sense.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 7 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Make sense.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Comment by cloud-fan** [comment]
@wangyum can you open a backport PR for 3.2?

**Comment by wangyum** [comment]
OK. PR for 3.2: https://github.com/apache/spark/pull/34627
